File: go/tests/faith_board_test.go
// faith_board_test.go
//
// Verifies the CircleAI.Faith port (faith_board.go): services in a window
// (ordered), recent prayers newest-first with limit, scripture lookup
// (case-sensitive tradition/book) and by-tradition listing (case-insensitive).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestFaith_ServicesAndPrayers(t *testing.T) {
	b := circleai.NewInMemoryFaithBoard()
	base := time.Date(2026, 7, 5, 8, 0, 0, 0, time.UTC)
	b.Schedule(circleai.FaithService{ServiceId: "s2", CommunityName: "Grace", Title: "Evening", StartUtc: base.Add(12 * time.Hour), Location: "Hall"})
	b.Schedule(circleai.FaithService{ServiceId: "s1", CommunityName: "Grace", Title: "Morning", StartUtc: base, Location: "Hall"})
	b.Schedule(circleai.FaithService{ServiceId: "s3", CommunityName: "Grace", Title: "NextWeek", StartUtc: base.AddDate(0, 0, 7), Location: "Hall"})

	svc := b.ServicesBetween(base, base.Add(24*time.Hour))
	if len(svc) != 2 || svc[0].ServiceId != "s1" || svc[1].ServiceId != "s2" {
		t.Fatalf("services-between ordered failed: %+v", svc)
	}

	now := time.Now().UTC()
	b.SubmitPrayer(circleai.PrayerRequest{RequestId: "p1", Author: "A", Body: "old", SubmittedUtc: now.Add(-time.Hour)})
	b.SubmitPrayer(circleai.PrayerRequest{RequestId: "p2", Author: "B", Body: "new", SubmittedUtc: now, IsAnonymous: true})
	rec := b.RecentPrayers(20)
	if len(rec) != 2 || rec[0].RequestId != "p2" {
		t.Fatalf("recent prayers newest-first failed: %+v", rec)
	}
	if lim := b.RecentPrayers(1); len(lim) != 1 || lim[0].RequestId != "p2" {
		t.Fatalf("recent prayers limit failed: %+v", lim)
	}
}

func TestFaith_Scripture(t *testing.T) {
	b := circleai.NewInMemoryFaithBoard()
	b.AddScripture(circleai.ScriptureReference{ReferenceId: "r2", Tradition: "Christian", Book: "John", Chapter: 3, Verse: 16, Text: "For God..."})
	b.AddScripture(circleai.ScriptureReference{ReferenceId: "r1", Tradition: "Christian", Book: "Psalms", Chapter: 23, Verse: 1, Text: "The Lord..."})

	if got, ok := b.Lookup("Christian", "John", 3, 16); !ok || got.ReferenceId != "r2" {
		t.Fatalf("scripture lookup = %+v ok=%v", got, ok)
	}
	// Tradition is case-sensitive per the C# == comparison.
	if _, ok := b.Lookup("christian", "John", 3, 16); ok {
		t.Fatalf("case-mismatched tradition must not match")
	}
	byTrad := b.ByTradition("CHRISTIAN")
	if len(byTrad) != 2 || byTrad[0].ReferenceId != "r1" || byTrad[1].ReferenceId != "r2" {
		t.Fatalf("by-tradition (case-insensitive, sorted) failed: %+v", byTrad)
	}
}

File: typescript/tests/faith_board.test.ts
// faith_board.test.ts
// Verifies the CircleAI.Faith port: services between, recent prayers, scripture
// lookup + by-tradition.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryFaithBoard,
  FaithDomainContext,
  faithService,
  prayerRequest,
  scriptureReference,
} from "../src/faith/index";

describe("InMemoryFaithBoard", () => {
  it("lists services within a window ascending", () => {
    const b = new InMemoryFaithBoard();
    b.schedule(faithService("s1", "Grace", "Sunday", new Date("2026-01-04T09:00:00Z"), "Main Hall"));
    b.schedule(faithService("s2", "Grace", "Midweek", new Date("2026-01-07T18:00:00Z"), "Room A"));
    b.schedule(faithService("s3", "Grace", "Later", new Date("2026-02-01T09:00:00Z"), "Main Hall"));
    assert.deepEqual(
      b
        .servicesBetween(new Date("2026-01-01T00:00:00Z"), new Date("2026-01-31T00:00:00Z"))
        .map((s) => s.serviceId),
      ["s1", "s2"],
    );
  });

  it("lists recent prayers newest-first capped by limit", () => {
    const b = new InMemoryFaithBoard();
    b.submitPrayer(prayerRequest("p1", "Ann", "one", new Date("2026-01-01T00:00:00Z"), false));
    b.submitPrayer(prayerRequest("p2", "", "anon", new Date("2026-01-03T00:00:00Z"), true));
    b.submitPrayer(prayerRequest("p3", "Cy", "three", new Date("2026-01-02T00:00:00Z"), false));
    assert.deepEqual(
      b.recentPrayers().map((p) => p.requestId),
      ["p2", "p3", "p1"],
    );
    assert.deepEqual(
      b.recentPrayers(1).map((p) => p.requestId),
      ["p2"],
    );
  });

  it("looks up scripture exactly and lists by tradition case-insensitively", () => {
    const b = new InMemoryFaithBoard();
    b.addScripture(scriptureReference("ref1", "Christian", "John", 3, 16, "For God so loved..."));
    b.addScripture(scriptureReference("ref2", "Christian", "Psalms", 23, 1, "The Lord is my shepherd..."));
    assert.equal(b.lookup("Christian", "John", 3, 16)?.referenceId, "ref1");
    assert.equal(b.lookup("Christian", "John", 3, 17), undefined);
    assert.deepEqual(
      b.byTradition("christian").map((r) => r.referenceId).sort(),
      ["ref1", "ref2"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(FaithDomainContext.systemPromptSnippet.includes("[DOMAIN: Faith]"));
    assert.deepEqual(FaithDomainContext.complianceFlags, ["POPIA", "Non_Denominational_Respect"]);
    assert.deepEqual(FaithDomainContext.suggestedTools, ["scripture_tools", "document_editor", "calendar"]);
  });
});
